distributed/shuffle/_merge.py

```diff
@@ -0,0 +1,447 @@
+
+import toolz
+
+from dask.base import is_dask_collection, tokenize
+from dask.core import keys_in_tasks
+from dask.highlevelgraph import HighLevelGraph
+from dask.layers import Layer
+from dask.utils import stringify, stringify_collection_keys
+
+from distributed.protocol.serialize import to_serialize
+from distributed.shuffle._shuffle import (
+    ShuffleId,
+    _get_worker_extension,
+    barrier_key,
+    shuffle_barrier,
+    shuffle_transfer,
+)
+
+if TYPE_CHECKING:
+    import pandas as pd
+    from pandas._typing import IndexLabel, MergeHow, Suffixes
+
+    from dask.dataframe.core import _Frame
+
+
+_HASH_COLUMN_NAME = "__hash_partition"
+
+
+def _prepare_index_for_partitioning(df: pd.DataFrame, index: IndexLabel):
+    import pandas as pd
+
+    from dask.dataframe.core import _Frame
+
+    list_like = pd.api.types.is_list_like(index) and not is_dask_collection(index)
+
+    if not isinstance(index, _Frame):
+        if list_like:
+            # Make sure we don't try to select with pd.Series/pd.Index
+            index = list(index)
+        index = df._select_columns_or_index(index)
+    elif hasattr(index, "to_frame"):
+        # If this is an index, we should still convert to a
+        # DataFrame. Otherwise, the hashed values of a column
+        # selection will not match (important when merging).
+        index = index.to_frame()
+    return index
+
+
+def _calculate_partitions(df: pd.DataFrame, index: IndexLabel, npartitions: int):
+    index = _prepare_index_for_partitioning(df, index)
+    from dask.dataframe.shuffle import partitioning_index
+
+    partitions = index.map_partitions(
+        partitioning_index,
+        npartitions=npartitions or df.npartitions,
+        meta=df._meta._constructor_sliced([0]),
+        transform_divisions=False,
+    )
+    df2 = df.assign(**{_HASH_COLUMN_NAME: partitions})
+    df2._meta.index.name = df._meta.index.name
+    return df2
+
+
+def hash_join_p2p(
+    lhs: _Frame,
+    left_on: IndexLabel | None,
+    rhs: _Frame,
+    right_on: IndexLabel | None,
+    how: MergeHow = "inner",
+    npartitions: int | None = None,
+    suffixes: Suffixes = ("_x", "_y"),
+    indicator: bool = False,
+):
+    from dask.dataframe.core import Index, new_dd_object
+
+    if npartitions is None:
+        npartitions = max(lhs.npartitions, rhs.npartitions)
+
+    if isinstance(left_on, Index):
+        left_on = None
+        left_index = True
+    else:
+        left_index = False
+
+    if isinstance(right_on, Index):
+        right_on = None
+        right_index = True
+    else:
+        right_index = False
+    # TODO: right/left_index still has to be forwarded
+    merge_kwargs = dict(
+        how=how,
+        left_on=left_on,
+        right_on=right_on,
+        left_index=left_index,
+        right_index=right_index,
+        suffixes=suffixes,
+        indicator=indicator,
+    )
+    # dummy result
+    # Avoid using dummy data for a collection it is empty
```

Comment:
I'm not sure I'm following, what do you mean here?

Reply:
copy paste